core/src/main/scala/scalaz/Cord.scala

```diff
@@ -1,163 +1,133 @@
-   * Prepends a `Char` to the beginning of this `Cord`.
-   * Time complexity: O(1)
-   */
-  def -:(x: Char): Cord = cord(String.valueOf(x) +: self)
-
-  /**
-   * Appends a `Char` to the end of this `Cord`.
-   * Time complexity: O(1)
-   */
-  def :-(x: => Char): Cord = cord(self :+ String.valueOf(x))
-
-  /**
-   * Removes the first character of this `Cord`.
-   * Time complexity: O(1)
-   */
-  def tail: Cord = drop(1)
-
-  /**
-   * Removes the last character of this `Cord`.
-   * Time complexity: O(1)
-   */
-  def init: Cord = take(self.measure - 1)
-
-  /**
-   * Removes the first `n` characters from the front of this `Cord`.
-   * Time complexity: O(min N (N - n))
-   */
-  def drop(n: Int): Cord = split(n)._2
-
-  /**
-   * Returns the first `n` characters at the front of this `Cord`.
-   * Time complexity: O(min N (N - n))
-   */
-  def take(n: Int): Cord = split(n)._1
-
-  /**
-   * Modifies each character in this `Cord` by the given function.
-   * Time complexity: O(N)
-   */
-  def map(f: Char => Char): Cord = cord(self map (_ map f))
-
-  def toList: List[Char] = toVector.toList
-  def toStream: Stream[Char] = toVector.toStream
-  def toVector: Vector[Char] = self.foldMap(_.toVector)
-  override def toString: String = {
-    val sb = new java.lang.StringBuilder(self.measure)
-    self.foreach { x =>
-      val _ = sb append x
-    }
+  /** Explicitly construct the output String represented by this Cord */
+  final def shows: String = {
+    val sb = new StringBuilder
+    Cord.unsafeAppendTo(this, sb)
     sb.toString
   }
 
-  /** Transforms each character to a `Cord` according to the given function and concatenates them all into one `Cord`. */
-  def flatMap(f: Char => Cord): Cord = toVector.foldLeft(Cord())((as, a) => as ++ f(a))
+  // Before adding any methods to this class, please consider if they can be
+  // reasonably implemented with future, optimised, representations of Cord. The
+  // purpose of this class is to generate Strings as fast as possible, not a
+  // general text manipulation structure. For example, a previous version used
+  // FingerTree, and many methods were added. But FingerTree was very
+  // inefficient at creating Strings yet we could not fix these performance
+  // problems because of the need for backcompatibility.
 
-  /** Returns whether this `Cord` will expand to an empty string. */
-  def isEmpty: Boolean = !nonEmpty
-
-  /** Returns whether this `Cord` will expand to a non-empty string. */
-  def nonEmpty: Boolean = self.iterator.exists(_.nonEmpty)
+  /** Strict evaluation variant of Monoid.append */
+  final def ++(o: Cord): Cord = Cord.Branch(this, o)
 }
-
 object Cord {
-  private def cord[A](v: FingerTree[Int, String]): Cord = new Cord(v)
-
-  implicit def stringToCord(s: String): Cord = cord(FingerTree.single[Int, String](s))
+  private[scalaz] final case class Leaf private (
+    s: String
+  ) extends Cord
+
+  private[scalaz] final case class Branch private (
+    leftDepth: Int,
```

Comment:
So terrible it's awesome! I love it. 😆

Reply:
I'd love some benchmarks on real data to see how often this hack ever gets invoked. Ideally it should only trigger < 0.001% of the time.